### PR TITLE
contracts/accountbook: implement accountbook contract

### PR DESCRIPTION
**Pull request by rjl493456442**
This PR implements the built-in payment channel which can be used in LES protocol

### CODE REVIEW

File: contracts/accountbook/contract/accountbook.sol [hunk 1]
```diff
@@ -0,0 +1,212 @@
+pragma solidity ^0.5.12;
```
Comment (gballet) [inline contracts/accountbook/contract/accountbook.sol]: You should also add the copyright notice.

File: contracts/accountbook/drawee.go [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+	}
+	// Drawer has a opened withdrawal request, no matter it passes the challenge
+	// period or not, minus this part.
+	if remaining.Cmp(req.Amount) < 0 {
+		return nil, ErrNotEnoughDeposit
+	}
+	return new(big.Int).Sub(remaining, req.Amount), nil
+}
+
+// Cash cashes all unpaid payment by given drawer.
```
Comment (gballet) [inline contracts/accountbook/drawee.go]: by a given drawer

File: contracts/accountbook/drawee.go [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+		_, err := bind.WaitMined(context, drawee.dBackend, tx)
+		if err != nil {
+			return err
+		}
+	}
+	log.Info("Cashed cheque", "drawer", drawer, "amount", unpaid, "cumulative", lastReceived.Amount)
+	return nil
+}
+
+// ListenWithdraw watches new withdraw event triggered by drawers.
```
Comment (gballet) [inline contracts/accountbook/drawee.go]: events

File: contracts/accountbook/drawee.go [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+func (drawee *ChequeDrawee) ListenWithdraw() (event.Subscription, chan *contract.AccountBookWithdrawEvent, error) {
+	sink := make(chan *contract.AccountBookWithdrawEvent)
+	sub, err := drawee.book.contract.WatchWithdrawEvent(nil, sink, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+	return sub, sink, nil
+}
+
+// ListCheques returns all cheques drawee received.
```
Comment (gballet) [inline contracts/accountbook/drawee.go]: that the drawee

File: contracts/accountbook/errors.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package accountbook
+
+import "errors"
+
+// ErrNotEnoughDeposit is returned if the cheque drawer doesn't have enough
+// balance to spend. Whenever drawee receives this error, it should emit a
```
Comment (gballet) [inline contracts/accountbook/errors.go]: the/a drawee

File: contracts/accountbook/errors.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+// cash operation as soon as possible.
+var ErrNotEnoughDeposit = errors.New("deposit is not enough")
+
+// StaleChequeError wraps a error msg and the evidence for a stale cheque.
```
Comment (gballet) [inline contracts/accountbook/errors.go]: an error message

File: les/client.go [hunk 7]
```diff
@@ -264,15 +274,42 @@ func (s *LightEthereum) Stop() error {
...
+			chequeSigner := func(data []byte) ([]byte, error) {
+				return wallet.SignData(account, accounts.MimetypeDataWithValidator, data)
+			}
+			channelManager, err := payment.NewPaymentChannelManager(payment.DefaultPaymentChannelDrawerConfig, s.chainReader, bind.NewRawTransactor(wallet.SignTx, account), chequeSigner, s.address, contract, deploy, s.paymentDb)
+			if err != nil {
+				log.Warn("Failed to setup cheque drawee", "error", err)
+				return
+			}
+			s.channelManager = channelManager
+			atomic.StoreUint32(&s.paymentInited, 1) // Mark payment channel is available now
```
Comment (gballet) [inline les/client.go]: as available now

**Comment by rjl493456442** [comment]
@gballet Really appreciate for your review! 

Since this payment channel is the one-to-one channel, the cost of the light client to setup routes is a little bit high. So I work on another payment channel called `lotteryBook`. You can check the work-in-progress branch here https://github.com/rjl493456442/go-ethereum/tree/lottery-book 

But again, really appreciate your time and precious feedback. I'll definitely fix these grammar issues in the PR to supersede this one!

**Comment by adamschmideg** [comment]
Closing in favor of #20544

**Pull request closed without merging**